design_notes/XDN03 - xlang Type System.md

```diff
@@ -15,3 +15,255 @@ status: draft
+of the type parameters have been fully specified. Non-parameterized interfaces may inherit from
+concrete interfaces, but they may not inherit from a parameterized interface.
+
+#### Parameterized Interface Examples
+
+For example, xlang's foundation types includes a set of parameterized interfaces to represent
+collections.
+
+- `IIterable<T>` represents a collection that can be enumerated
+- `IVector<T>` represents a collection that can be randomly accessed
+- `IMap<K,V>` represents a collection that maps from keys to values
+
+`IIterable<T>` is a parameterized interface. `IIterable<String>` is a concrete interface, as all
+of the type parameters have been specified. 
+
+Both `IVector<T>` and `IMap<K,V>` can be enumerated. `IVector<T>` derives from `IIterable<T>`.
+In this case, the same type parameter is used for both `IVector` and `IIterable`. However,
+`IMap<K,V>` derives from `IIterable<IKeyValuePair<K,V>>`. In this case, the `T` type parameter of
+`IIterable` is instantiated as parameterized type `IKeyValuePair<K,V>`, using the type parameters
+defined as part of of `IMap<K,V>`.
+
+Using the same collection interfaces defined above, you could define an non parameterized interface
+`IPropertySet` that inherits from `IMap<String, Object>`. Note the use of concrete types in the
+declaration of IMap, where in the previous example we used type parameters.
+
+## Delegates
+
+Delegates are xlang types that act as a type-safe function pointer. They are similar conceptually
+to an [interface](#interfaces) that declares one and only one method. Defining delegates as a
+separate type from interfaces enables better integration with free functions and/or [anonymous
+functions](https://en.wikipedia.org/wiki/Anonymous_function) in language projections. Delegates are
+often (but not exclusively) used to declare the signature for xlang event handlers.
+
+xlang delegates are named types and define a method signature. Delegate method signatures follow the
+same rules for parameters as interface methods do.
+
+Like interfaces, delegates require a GUID identifier. THis identifier can be provided explicitly or
+generated implicitly from the type's name.
+
+Like interfaces, delegates can be parameterized. Parameterized delegates may declare arguments
+and/or their return value type in terms of the type parameters instead of as concrete types.
+
+Delegates are not versionable. Once they have been defined, they may never be changed.
+
+## Classes
+
+Classes are xlang types that expose behavior and encapsulate data. Often times, classes can be activated
+
+Classes are additively versionable. Subsequent versions of a given enum may add new members. 
+Pre-existing members may not be removed or modified in any way.
+
+## Object Members
+
+### Methods
+
+#### Method Overloading
+
+#### Parameters
+
+##### Array Parameters
+
+### Properties
+
+### Events
+
+### Intrinsic Operations
```

Comment:
I'm still wondering if we need an intrinsic operation for "associated objects" or some other way that the source object can actually know something about its projection(s). In Project Rome there is a current dilemma where Java Event Handlers require weak references because at the JNI layer a global reference is taken to store the event handler as a native reference in the source object. The java garbage collector can't see through this global reference and no clean up can happen until unsubscribe is called. I believe Cpp\WinRT takes the alternate approach and always attaches the Event Handler as a weak reference but that required us manually take strong references to objects since the projection object we used was somewhat ephemeral anyway. Basically we need to start to think about how we are going to cache and reuse projections and if any of that is best served on the source object itself such that it can cooperate with the projection.

Reply:
WinRT delegates don't support weak references today.

Reply:
Am I understanding this to further support the need of cooperative ref counting / reference holding between the projection and its source?

Reply:
THis issue is now being tracked by #144